// angular-estree-parser/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const context_1 = require("./context");
const transform_1 = require("./transform");
const transform_microsyntax_1 = require("./transform-microsyntax");
const utils_1 = require("./utils");
function parse(input, parseNg) {
    const { ast: rawNgAst, comments } = parseNg(input);
    const context = new context_1.Context(input);
    const _t = (n) => transform_1.transform(n, context);
    const ast = _t(rawNgAst);
    ast.comments = comments.map(_t);
    return ast;
}
function parseBinding(input) {
    return parse(input, utils_1.parseNgBinding);
}
exports.parseBinding = parseBinding;
function parseSimpleBinding(input) {
    return parse(input, utils_1.parseNgSimpleBinding);
}
exports.parseSimpleBinding = parseSimpleBinding;
function parseInterpolation(input) {
    return parse(input, utils_1.parseNgInterpolation);
}
exports.parseInterpolation = parseInterpolation;
function parseAction(input) {
    return parse(input, utils_1.parseNgAction);
}
exports.parseAction = parseAction;
function parseTemplateBindings(input) {
    return transform_microsyntax_1.transformTemplateBindings(utils_1.parseNgTemplateBindings(input), new context_1.Context(input));
}
exports.parseTemplateBindings = parseTemplateBindings;
